articles/budougumi0617.re

```diff
@@ -1,18 +1,480 @@
+
+//footnote[tog][@<href>{https://tour.golang.org}]
+
+=== @<code>{golang.org/x/website}パッケージ
+ * @<href>{https://godoc.org/golang.org/x/website}
+ * @<href>{https://github.com/golang/website}
+
+Goの公式サイトは御存知のとおり、@<tt>{golang.org}@<fn>{golangorg}にあります。
+公式サイトのHOME画面を管理しているのが@<code>{website}パッケージです。
+公式サイトのその他のページはGo本体のリポジトリに含まれていたりします@<fn>{go_doc}。
+Google Cloudにある@<i>{google.golang.org パッケージの一覧}@<fn>{googlegolangorg}から、各GoDocへリダイレクトする簡易Webサーバなどの実装@<fn>{google_pkg_list}も同梱されています。
+
+
+//footnote[go_doc][@<href>{https://github.com/golang/go/tree/go1.13/doc}]
+//footnote[golangorg][@<href>{https://golang.org}]
+//footnote[google_pkg_list][@<href>{https://cloud.google.com/go/google.golang.org}]
+//footnote[googlegolangorg][@<href>{https://github.com/golang/website/tree/master/cmd/googlegolangorg}]
+
+
+=== @<code>{golang.org/x/xerrors}パッケージ
+ * @<href>{https://godoc.org/golang.org/x/xerrors}
+ * @<href>{https://github.com/golang/xerrors}
+
+Goでスタックトレースを含んだエラー処理を行ないたいならば、@<code>{github.com/pkg/errors}パッケージを使うのがデファクトスタンダードでした。
+@<code>{xerrors}パッケージはGo2に向けて提案された@<i>{Proposal: Go 2 Error Inspection}@<fn>{go2_error}をGo1向けに実装したライブラリです。
+@<code>{xerrors}パッケージの@<code>{Newf}（あるいは@<code>{Errof}）関数から生成されたエラーは内部にスタックトレースを持ちます。このスタックトレースはプリントフォーマットで@<code>{%+v} verbeを使って出力することができます。
+また、@<code>{xerrors}パッケージはスタックトレース以外にもエラーの同値性を検証する@<code>{Is}関数、取得したエラーから具体的な型のオブジェクトを抽出できる@<code>{As}関数が提供されています。
+
+#@# textlint-disable
+
+//list[xerr_sample][xerrorsパッケージを使ったエラーハンドリング@<fn>{play_xerr}][go]{
+package main
+
+import (
+  "fmt"
+
+  "golang.org/x/xerrors"
+)
+
+type MyError struct {
+  body string
+}
+
+func (_ *MyError) Error() string {
+  return "original error"
+}
+
+// 通常のエラーには存在しないメソッド
+func (me *MyError) Body() string {
+  return me.body
+}
+
+func main() {
+  base := &MyError{"original body"}
+  err := xerrors.Errorf("wraped: %w", base)
+  fmt.Printf("%+v\n", err)
+  // errの中にMyError型が含まれていればtrue
+  fmt.Println("xerrors.Is", xerrors.Is(err, base))
+
+  var me *MyError
+  // errの中からMyError型を抽出できたらmeに代入する
+  if xerrors.As(err, &me) {
+    fmt.Printf("body: %q\n", me.Body())
+
+  }
+}
+//}
+
+#@# textlint-enable
+
+@<list>{xerr_sample}を実行した結果は次のとおりです。スタックトレースからファイル名と行数が取得できています。
+@<code>{Is}関数では、ネストしたエラーの型もチェックすることができます。
+また、@<code>{As}関数によってエラーの中に含まれている独自型の情報も取得することができました。
+
+//cmd{
+wraped:
+    main.main
+        /tmp/sandbox096453308/prog.go:24
+  - original error
+xerrors.Is true
+body: "original body"
+//}
+
+//footnote[play_xerr][@<href>{https://play.golang.org/p/9Vq2jTUiL5b}]
+
+このような機能を含む@<code>{xerrors}パッケージですが、2019年9月3日に公開されたGo1.13で@<code>{xerrors}パッケージで定義された関数が公式パッケージの@<code>{errors}パッケージに正式に導入されました。
+しかし、@<code>{%+w}や@<code>{%+v}によるスタックトレースの表示の採用は見送られています@<fn>{xerr_frame}。
+
+//footnote[xerr_frame][Go1.13のエラーオブジェクトは内部にスタックトレース（@<code>{Frame}）情報を持っていない]
+
+Go1.13の公式パッケージの@<code>{gerrors}パッケージでは依然としてスタックトレースを取得することができないため、スタックトレースの表示が必要な場合は@<code>{xerrors}パッケージを利用して、不要な場合には標準ライブラリの@<code>{errors}パッケージを利用してください。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 410 sentence length(129) exceeds the maximum sentence length of 100.
Over 29 characters. (ja-technical-writing/sentence-length)